app/models/manageiq/providers/openshift/inventory/collector/container_manager.rb

```diff
@@ -1,57 +1,66 @@
 class ManageIQ::Providers::Openshift::Inventory::Collector::ContainerManager < ManageIQ::Providers::Kubernetes::Inventory::Collector::ContainerManager
-  attr_reader :version
+  def clusterversion
+    @clusterversion ||= begin
+      openshift_connection("config.openshift.io/v1").get_cluster_version("version") if openshift_version == "v4"
+    end
+  end
 
   def routes
-    @routes ||= fetch_entity(openshift_connection("route.openshift.io"), "routes")
+    @routes ||= fetch_entity(openshift_connection("route.openshift.io/v1"), "routes")
   end
 
   def namespaces_by_name
     @namespaces_by_name ||= namespaces.index_by { |ns| ns.metadata.name }
   end
 
   def projects
-    @projects ||= fetch_entity(openshift_connection("project.openshift.io"), "projects")
+    @projects ||= fetch_entity(openshift_connection("project.openshift.io/v1"), "projects")
   end
 
   def build_configs
-    @build_configs ||= fetch_entity(openshift_connection("build.openshift.io"), "build_configs")
+    @build_configs ||= fetch_entity(openshift_connection("build.openshift.io/v1"), "build_configs")
   end
 
   def builds
-    @builds ||= fetch_entity(openshift_connection("build.openshift.io"), "builds")
+    @builds ||= fetch_entity(openshift_connection("build.openshift.io/v1"), "builds")
   end
 
   def templates
-    @templates ||= fetch_entity(openshift_connection("template.openshift.io"), "templates")
+    @templates ||= fetch_entity(openshift_connection("template.openshift.io/v1"), "templates")
   end
 
   def images
-    @images ||= refresher_options.get_container_images ? fetch_entity(openshift_connection("image.openshift.io"), "images") : {}
+    @images ||= refresher_options.get_container_images ? fetch_entity(openshift_connection("image.openshift.io/v1"), "images") : {}
   end
 
   private
 
   def openshift_connection(group)
-    detect_openshift_version! if version.nil?
-    send("openshift_connection_#{version}", group)
+    send("openshift_connection_#{openshift_version}", group)
+  end
+
+  def openshift_version
+    @openshift_version ||= detect_openshift_version!
   end
 
   def detect_openshift_version!
-    @version = begin
+    version = begin
       openshift_connection_v3
       "v3"
     rescue Kubeclient::ResourceNotFoundError
       nil
     end
 
-    @version ||= begin
-      openshift_connection_v4("project.openshift.io")
+    version ||= begin
+      openshift_connection_v4("project.openshift.io/v1")
       "v4"
     rescue Kubeclient::ResourceNotFoundError
       nil
     end
 
-    raise "Failed to detect OpenShift version" if @version.nil?
+    raise "Failed to detect OpenShift version" if version.nil?
+
```

Comment:
[cosmetic] Now that this method doesn't save to an instance variable, it could be simpler to read with `version =` and `||=`.  Do `return "v3"` and `return "v4"`, and here just raise unconditionally.

Reply:
I meant "without"